Adding Fragment Cache to AMS

joaomdmoura opened the pull request:
It's an upgrade based on the new Cache implementation #693.
Referenced: #802 

It allows to use the Rails conventions to cache specific attributes or associations, as you will see in the example ahead. 

It's optimized to re-use the cache in ever way as possible, in order to do this, the cached data will shared and resued on different methods, also the Serializers that are cached will be used when referenced as relationship at other serializer.

If there is some API endpoint that shouldn't be fully cached, you can still optimise it, using Fragment Cache on the attributes and relationships that you want to cache.

You can define the attributes by using ```only``` or ```except``` option on cache mehtod.

Example:

```ruby
class PostSerializer < ActiveModel::Serializer
  cache key: 'post', expires_in: 3.hours, only: [:title]
  attributes :title, :body

  has_many :comments

  url :post
end
```

It also work when defining a key to an attribute, in this case you should use the original name of the attribute (no worries, the key still will work):

```ruby
AlternateBlogSerializer = Class.new(ActiveModel::Serializer) do
  cache only: [:name]
  attribute :id
  attribute :name, key: :title
end
```

joaomdmoura commented:
@kurko and it's alive ;) made with :heart: 
Tests passing, Commits Squashed :+1:

kurko commented on README.md:
````diff
@@ -294,6 +295,25 @@ On this example every ```Post``` object will be cached with
 
+### Fragmenting Caching
+
+If there is some API endpoint that shouldn't be fully cached, you can still optmize it, using Fragment Cache on the attributes and relationships that you want to cache.
````
Typo, `optime`.

kurko commented on README.md:
````diff
@@ -294,6 +295,25 @@ On this example every ```Post``` object will be cached with
+
+If there is some API endpoint that shouldn't be fully cached, you can still optmize it, using Fragment Cache on the attributes and relationships that you want to cache.
+
+You can define the attribute or relationships by using ```only``` or ```except``` option on cache mehtod.
````
Typo, `mehtod`.

kurko commented on lib/active_model/serializer/adapter.rb:
```diff
@@ -55,15 +55,97 @@ def include_meta(json)
+          serializers[:non_cached].constantize.attribute(attribute, options)
+        end
+        fragment_associations(serializers, associations)
+      end
```
Could you extract all of these methods into another class? `Cache` perhaps? e.g

```
def cached_hash_root(cached_hash, non_cached_hash)
  cache.hash_root(cached_hash, non_cached_hash)
end

# ...

private

def cache
  Cache.new(self)
end
```

Otherwise, it's going to be hard to be maintained.

joaomdmoura replied:
Totally agree. I'll extract it to an standalone class and include it in the Adapter. If you have another suggestions just let me know

joaomdmoura replied:
Totally agree. I'll extract it to an standalone class and include it in the Adapter. If you have another suggestions just let me know

guilleiguaran commented:
/cc @spastorino

joaomdmoura commented:
@kurko, Updated, thank you for the review! Just let me know if you find anything else.

joaomdmoura commented:
@kurko, Updated, thank you for the review! Just let me know if you find anything else.

kurko commented on lib/active_model/serializer/adapter.rb:
```diff
@@ -1,7 +1,10 @@
   class Serializer
     class Adapter
       extend ActiveSupport::Autoload
+      include ActiveModel::Serializer::Cache
```
So, my suggestion was to really move the code out of the former class. Instead, you just moved it to a new file, but the code is still inside the Adapter.

In my opinion, the main problem with this approach is that you can't unit test the code you added, but only have integration tests, which aren't really good for refactoring the code later. My suggestion was to use composition, or in other words, call the cache class such as:

```ruby
def cached_object
   CachedResource.new(serializer).load
end
```

With that, you can unit test `CachedResource` separately. Besides that, the adapter has a huge amount of methods doing things that are not the adapter's responsibility.

What do you think? cc @guilleiguaran

joaomdmoura replied:
Got it. Indeed. Serialization_test is getting too big.

> Besides that, the adapter has a huge amount of methods doing things that are not the adapter's responsibility.

Convinced me

joaomdmoura replied:
@kurko, to avoid performance loss I'll just remove the fragment cache logic, that correspond for 90% of all code, agreed?:
```ruby
ef cached_object
         klass = serializer.class
        if klass._cache && !klass._cache_only && !klass._cache_except
          _cache_key = (klass._cache_key) ? "#{klass._cache_key}/#{serializer.object.id}-#{serializer.object.updated_at}" : serializer.object.cache_key
          klass._cache.fetch(_cache_key, klass._cache_options) do
            yield
          end

        elsif klass._cache_only && !klass._cache_except || !klass._cache_only && klass._cache_except
          FragmentCache.new(self, serializer).fetch
        else
          yield
        end
      end
```

kurko replied:
Sounds good, that's a good start. I don't think, though, that another class will result in any significant loss.

***

On another not, I think we need to name things better here; we need to better group the code. For instance, I have no idea from a glance what the following is:

```ruby
klass._cache_only && !klass._cache_except || !klass._cache_only && klass._cache_except
```

I'd rather have


```ruby
def cached_object
  # ...
  elsif fragment_cache?
      FragmentCache.new(self, serializer, @options, @root).fetch
  else
  #...
end

private

def fragment_cache?
  klass._cache_only && !klass._cache_except || !klass._cache_only && klass._cache_except
end
```

joaomdmoura replied:
The loss would be result of passing the block correspondent to the yield to order function. But yeah it's nothing big.

---------------------------------

Agreed. Even ```cached_object`` isn't a good name, I'm working on this too.

JoshSmith commented:
@joaomdmoura what can I do to help here?

thibaudgg commented:
@joaomdmoura looks awesome. :heart_eyes: 
One question, will [`fetch_multi`](http://api.rubyonrails.org/classes/ActiveSupport/Cache/Store.html#method-i-fetch_multi) be supported in the future? It would be really useful (and faster!) for fetching collection or association from cache in one read. :dash: 
Thanks!

joaomdmoura commented:
@JoshSmith tks man, but there are just some details missing, I'm working on it already :smile: but there are some improvements that I would like to work on in the near future that you can focus too, like the @thibaudgg suggestion.

@thibaudgg Tks! Unfortunately the new cache feature does not use ```fetch_multi```. But it's a great suggestion! It's more related with my last PR #693 adding the new cache feature, if you read this PR you will realize that I cache each object separately. Indeed change the way the ```Adapter``` works when retrieving an cached collection to use ```fetch_multi``` could be an improvement, maybe also a great opportunity to start benchmarking AMS features.

thibaudgg commented:
@joaomdmoura `fetch_multi` support would be a big plus against `jbuilder` and `jbuilder_cache_multi`! Thanks.

joaomdmoura commented:
@kurko, It's done.
- Extracted the fragment logic to a new class ```FragmentCache```
- Refactored some code from #693 
- Wrote some tests related to ```FragmentCache```

JoshSmith commented:
:beers:

kurko commented:
Great work here, @joaomdmoura! Before merging, I want to spend more time with it. I'll deploy it to our staging env and see how it behaves and whether I can catch bugs or not. I'd advise you doing the same over there :D

kurko commented on lib/active_model/serializer/fragment_cache.rb:
```diff
@@ -0,0 +1,99 @@
+
+        serializers = {cached: cached, non_cached: non_cached}
+        cached_attributes_and_association(klass, serializers)
+        return serializers
```
There's no need for this `return`.

joaomdmoura replied:
Yeah, I know, I just felt that I should make it explicitly.
Never mind I'm removing it. :stuck_out_tongue_closed_eyes:

kurko commented on lib/active_model/serializer/fragment_cache.rb:
```diff
@@ -0,0 +1,99 @@
+      def fragment_associations(serializers, associations)
+        associations.each do |association|
+          options = ",#{association[1][:association_options]}" if association[1].include?(:association_options)
+          eval("#{serializers[:non_cached]}.#{association[1][:type].to_s}(:#{association[0]}#{options})")
```
Could you clarify what's going on here? A comment in this case would be helpful, I think, as I can't really understand what's happening.

This `eval` ringed a lot of alarm sounds in my head. What is it doing? Is it really needed?

joaomdmoura replied:
Commenting. No there is no need, I just forgot it there while I was reviewing the code, I should have replaced it with ```.send```

guilleiguaran replied:
public_send please :)

kurko commented on lib/active_model/serializer/fragment_cache.rb:
```diff
@@ -0,0 +1,99 @@
+        if @root
+          hash[@root] = (core_cached[1]) ? core_cached[1].merge(core_non_cached[1]) : core_non_cached[1]
+        else
+          hash = (core_cached[1]) ? core_cached[1].merge(core_non_cached[1]) : core_non_cached[1]
```
This line is exactly the same as the one above. Could if be something like:

```ruby
cached_resource = (core_cached[1]) ? core_cached[1].merge(core_non_cached[1]) : core_non_cached[1]
hash = if @root
         { @root => cached_resource }
       else
         cached_resource
       end
```

Which allows you to remove the line 34.

joaomdmoura replied:
Agreed, but I transformed it into a ternary. Fewer lines, still understandable. 
```hash = (@root) ? { @root => cached_resource } : cached_resource```

kurko commented on lib/active_model/serializer/fragment_cache.rb:
```diff
@@ -0,0 +1,99 @@
+        non_cached_hash = @adapter.class.new(non_cached_serializer, @options).serializable_hash
+
+        if @serializer.root && @adapter.class == ActiveModel::Serializer::Adapter::JsonApi
+          cached_hash_root(cached_hash, non_cached_hash)
```
If you're adding code specific for `JsonApi`, I think you should push that into the adapter. You could use polymorphism.

You could create `ActiveModel::Serializer::Adapter::FragmentCache` and `ActiveModel::Serializer::Adapter::JsonApi::FragmentCache`. Then here you could just do:

```ruby
adapter.fragment_cache(cached_hash, non_cached_hash)
```

This would also get rid of `cached_hash_root` from this class, specially given it's `JsonApi` specific.

joaomdmoura replied:
Yeah, you are definitely right, there is no sense to have a method to a specific adapter using ```FragmentCache``` abstract as it is.
This is the only change I haven't worked yet, I'll give it some thought yet. But polymorphism indeed seems a good way to go.

kurko replied:
Ping me if you need help with it.

kurko commented on lib/active_model/serializer/fragment_cache.rb:
```diff
@@ -0,0 +1,99 @@
+
+      private
+
+      def cached_hash_root(cached_hash, non_cached_hash)
```
Given this method is JsonApi specific, we'd probably have to add a new method for each adapter. That's yet another argument for the polymorphic approach I mentioned above.

Also, it should be named something like `json_api_cache_root` or something like that.

kurko commented on test/adapter/json_api/linked_test.rb:
```diff
@@ -44,8 +45,8 @@ def test_include_multiple_posts_and_linked
-                           { title: "Hello!!", body: "Hello, world!!", id: "1", links: { comments: ['1', '2'], author: "1" } },
-                           { title: "New Post", body: "Body", id: "2", links: { comments: [], :author => "2" } }
+                           { :id=>"1", :title=>"Hello!!", :body=>"Hello, world!!", :links=>{:comments=>["1", "2"], :blog=>"999", :author=>"1"} },
+                           { :id=>"2", :title=>"New Post", :body=>"Body", :links=>{:comments=>[], :blog=>"999", :author=>"2"} }
```
Can we get rid of the hash-rocket version just for consistency? We have it differently everywhere. Also, it feels so Ruby 1.9 :grin:

joaomdmoura commented on lib/active_model/serializer/adapter.rb:
```diff
@@ -32,8 +35,36 @@ def self.adapter_class(adapter)
         "ActiveModel::Serializer::Adapter::#{adapter.to_s.classify}".safe_constantize
       end
 
+      def fragment_cache(*args)
```
@kurko abstracted method. Duck typing strategy.

joaomdmoura commented on lib/active_model/serializer/adapter.rb:
```diff
@@ -32,8 +35,36 @@ def self.adapter_class(adapter)
+
       private
 
+      def cache_check
```
Updated the old ```cache_object``` to ```cache_check```.
Also abstracted some of its behaviours into other methods. (bellow)

joaomdmoura commented on lib/active_model/serializer/adapter/fragment_cache.rb:
```diff
@@ -0,0 +1,90 @@
+module ActiveModel
```
I commented everything that could make it easy to support.

joaomdmoura commented on lib/active_model/serializer/adapter/fragment_cache.rb:
```diff
@@ -0,0 +1,90 @@
+          non_cached_hash = @adapter.class.new(non_cached_serializer, @options).serializable_hash
+
+          # Merge both results
+          @adapter.fragment_cache(cached_hash, non_cached_hash)
```
The duck typing implementation that was defined at ```adapter.rb``` similar to ```serializable_hash```

joaomdmoura commented on lib/active_model/serializer/adapter/fragment_cache.rb:
```diff
@@ -0,0 +1,90 @@
+
+          # Add cached associations to cached Serializer
+          cached_associations.each do |k,v|
+            serializers[:cached].constantize.public_send(v[:type], k, v[:association_options])
```
Using public_send @guilleiguaran :smile: thank you for point it out!

joaomdmoura commented:
Rebased with :heart: PR fully updated.

kurko commented:
Talking to @joaomdmoura, we decided that it's better if we start by making the cache separate by resource. Right now, it's caching the entire response, including associations.

Say we have `posts` with `comments` and you want to update `comments`. Currently, you'll have to manually invalidate the respective `post` cache. You'll have to write code outside AMS to invalidate that. We want to start simple.

With the new decision, the main resource (`post`) will be cached separately from each `comment`. Then, when a comment is updated, AMS should load the new cache automatically. That will mean we'll have to do some processing to fetch caches, but will be simpler to start with.

Although the current way is faster, it'll be much more complicated for end users. We'll start with the simpler approach, which should be fast enough for most applications, and then move to the integral cache, which will require us to put some effort in better documenting it.

thibaudgg commented:
@kurko sounds great!

joaomdmoura commented:
Finally. It's done! @kurko 
Updated the implementation as we have being discussing, added some new tests too.
It's rebased and squashed :smile:

JoshSmith commented:
@joaomdmoura  :muscle: :heart_eyes:

thibaudgg commented:
:+1:

brancusi commented:
Nice!

kurko commented:
I tried this in an app and couldn't make it work. Basically, when I add `cache only: [:name]` to my serializer, I get this:

```
  1) Error:
UsersTest::GET /users/{id}#test_0001_returns a user by id:
NoMethodError: undefined method `my_virtual_attribute' for #<User:0x0000011126d1c8>
    app/controllers/users_controller.rb:6:in `show'
    test/test_helper.rb:50:in `get'
    my_app_test_path_test.rb:346:in `block (4 levels) in <class:UsersTest>'
```

My serializer is like this:

```
class UserSerializer < ActiveModel::Serializer
  cache only: [ :name ]

  attributes :id,
             :name,
             :my_virtual_attribute

  def my_virtual_attribute
    :some_custom_value
  end
end
```

If I leave `cache` out, it works fine. I tested against f7fc56f05184. For your reference, the PR that adds support for virtual associations (not attributes) is https://github.com/rails-api/active_model_serializers/pull/792.

When fixing, please make sure that both virtual attributes and associations are working.

Great work, by the way.

kurko commented:
Needs to be rebased 😬

joaomdmoura commented:
@kurko fixed, new tests written, rebased and squashed. :smile:

kurko commented:
FYI, I'm trying to backport this PR into a fork that supports RC1 so I can test this in my app. If it works there, I'll merge it. This is so central to AMS; it has to work perfectly.

kurko commented on lib/active_model/serializer/adapter.rb:
```diff
@@ -32,8 +34,38 @@ def self.adapter_class(adapter)
       private
 
+      def cache_check(serializer)
+        @serializer = serializer
```
This mutation is breaking a test in https://github.com/ride/active_model_serializers/blob/master-rc1-with-fragment-cache/test/action_controller/json_api_linked_test.rb#L146-L150 (it's a branch which I'm backporting this into RC1 because that's what my company is using).

Basically, halfway through the serialization (within a execution of `serializable_hash`), `@serializer`, which was `PostSerializer`, becomes `AuthorSerializer` after this. That same test is passing in this PR, so I'm not really sure what's going on, but regardless of that, mutating a variable that's used to figure out how to serialize stuff is bad and can lead to nasty behaviors in the future. If we're going to mutate, we need to simply instantiate a new object then or make sure the serialization has finished.

@joaomdmoura if you're available, let's discuss this over chat :)

kurko commented:
@joaomdmoura first of all, great work here! This got better and better every iterations. I tried this in my app and tests are green :grin: 

However, I need to show you something. I'm probably doing something wrong, but I ran some tests and got this:

```
# app/serializers/users.rb
class UserSerializer < ActiveModel::Serializers
  cache only: [
      :name, :email, :primary_role, :secondary_roles, 
      :mobile_number, :date_of_birth
    ]

  # ...
end

# test/integration/users_test.rb
Benchmark.bm do |x|
  x.report do
    1000.times do
      get "/users/#{user.id}", nil, oauth_headers(:route_matching)
    end
  end
end

Results:

               user        system    total     real
no cache       21.550000   1.820000  23.370000 ( 28.894494)
cache          22.270000   1.910000  24.180000 ( 30.448672)
fragment cache 32.320000   2.210000  34.530000 ( 43.051876)

(cache means `only: []` wasn't used in the serializer)
```

Do you have any idea why this is?

joaomdmoura commented:
Great @kurko 
I'm really glad we pulled it off.
The benchmark issue ( #832 ) now would be awesome, btw I'm already working on it.
The cache obviously improved overall performance. 

I'm sad about the fragment cache result, but as we discussed it would be really useful for attributes that involves some kind of logic, as virtual attributes on the serializer itself. Would be great to benchmark this case too :smile: 

But there is some improvement we can work on to optimize fragment cache speed, once benchmark tests are done will be way easier to keep track of it.

kurko commented:
Awesome. I'm testing it and if I don't find issues in a day or two, I'll merge this :smile:

Catalectic commented:
Any update on this? Really looking forward to this :)

kurko commented:
So, does any one have any problems with this? If not, I'm merging this pretty soon (likely tomorrow).

guilleiguaran commented:
@kurko can you paste the code you used for benchmarks in a gist?

kurko commented:
```ruby
# both fragment
cache only: [:field1, field2, :etc]

# and not fragment
cache

# then, in an integration test:
Benchmark.bm do |x|
  x.report do
    1000.times do
      get "/users/#{user.id}", nil, oauth_headers(:route_matching)
    end
  end
end
```

joaomdmoura commented:
\o/ :laughing:

mib32 commented on lib/active_model/serializer/adapter/fragment_cache.rb:
```diff
@@ -0,0 +1,78 @@
+        end
+
+        def fragment_serializer(name, klass)
+          cached     = "#{name.capitalize}CachedSerializer"
```
If different serializers share the same model, but has totally different attributes, would their caches collide, even if the key is different? I believe i see it on my machine. If you believe this could happen, i can present those tests to you, but i need time to set it up.

maybe it should go something like "#{serializer.name}CachedSerializer"?

joaomdmoura replied:
Yeap, I think it could happen. Not the cache_key itself, because it takes the file path into account but the Class name would colide.
I'll check it on the master codebase, create a new issue for that and update it.

mib32 replied:
@joaomdmoura thanks, that would be great! 
Here is the stack trace of the error i've got, when the classes collide, in case you would like to look at it. The latter called serializer tries to look for an attribute that was held by the first serializer, but not in the latter. Am i abusing the way serializers should be used, if i'm the first one who's got that error?

```
Failure/Error: get '/markers/local.json?coordinates=55.75,36.00&limit=10'
     NoMethodError:
       undefined method `name' for #<MarkerDistanceSerializer:0x007f80339b6eb8>
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/bundler/gems/active_model_serializers-e7d3323d2352/lib/active_model/serializer.rb:155:in `public_send'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/bundler/gems/active_model_serializers-e7d3323d2352/lib/active_model/serializer.rb:155:in `block in attributes'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/bundler/gems/active_model_serializers-e7d3323d2352/lib/active_model/serializer.rb:151:in `each'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/bundler/gems/active_model_serializers-e7d3323d2352/lib/active_model/serializer.rb:151:in `each_with_object'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/bundler/gems/active_model_serializers-e7d3323d2352/lib/active_model/serializer.rb:151:in `attributes'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/bundler/gems/active_model_serializers-e7d3323d2352/lib/active_model/serializer/adapter/json.rb:15:in `block in serializable_hash'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/bundler/gems/active_model_serializers-e7d3323d2352/lib/active_model/serializer/adapter.rb:51:in `block in cache_check'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/gems/activesupport-4.2.3/lib/active_support/cache.rb:299:in `block in fetch'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/gems/activesupport-4.2.3/lib/active_support/cache.rb:585:in `block in save_block_result_to_cache'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/gems/activesupport-4.2.3/lib/active_support/cache.rb:547:in `block in instrument'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/gems/activesupport-4.2.3/lib/active_support/notifications.rb:166:in `instrument'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/gems/activesupport-4.2.3/lib/active_support/cache.rb:547:in `instrument'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/gems/activesupport-4.2.3/lib/active_support/cache.rb:584:in `save_block_result_to_cache'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/gems/activesupport-4.2.3/lib/active_support/cache.rb:299:in `fetch'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/bundler/gems/active_model_serializers-e7d3323d2352/lib/active_model/serializer/adapter.rb:50:in `cache_check'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/bundler/gems/active_model_serializers-e7d3323d2352/lib/active_model/serializer/adapter/json.rb:14:in `serializable_hash'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/bundler/gems/active_model_serializers-e7d3323d2352/lib/active_model/serializer/adapter/flatten_json.rb:6:in `serializable_hash'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/bundler/gems/active_model_serializers-e7d3323d2352/lib/active_model/serializer/adapter/fragment_cache.rb:27:in `fetch'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/bundler/gems/active_model_serializers-e7d3323d2352/lib/active_model/serializer/adapter.rb:54:in `cache_check'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/bundler/gems/active_model_serializers-e7d3323d2352/lib/active_model/serializer/adapter/json.rb:14:in `serializable_hash'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/bundler/gems/active_model_serializers-e7d3323d2352/lib/active_model/serializer/adapter/flatten_json.rb:6:in `serializable_hash'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/bundler/gems/active_model_serializers-e7d3323d2352/lib/active_model/serializer/adapter/json.rb:10:in `block in serializable_hash'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/bundler/gems/active_model_serializers-e7d3323d2352/lib/active_model/serializer/array_serializer.rb:6:in `each'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/bundler/gems/active_model_serializers-e7d3323d2352/lib/active_model/serializer/array_serializer.rb:6:in `each'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/bundler/gems/active_model_serializers-e7d3323d2352/lib/active_model/serializer/adapter/json.rb:10:in `map'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/bundler/gems/active_model_serializers-e7d3323d2352/lib/active_model/serializer/adapter/json.rb:10:in `serializable_hash'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/bundler/gems/active_model_serializers-e7d3323d2352/lib/active_model/serializer/adapter/flatten_json.rb:6:in `serializable_hash'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/bundler/gems/active_model_serializers-e7d3323d2352/lib/active_model/serializer/adapter.rb:24:in `as_json'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/gems/activesupport-4.2.3/lib/active_support/json/encoding.rb:35:in `encode'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/gems/activesupport-4.2.3/lib/active_support/json/encoding.rb:22:in `encode'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/gems/activesupport-4.2.3/lib/active_support/core_ext/object/json.rb:37:in `to_json_with_active_support_encoder'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/gems/actionpack-4.2.3/lib/action_controller/metal/renderers.rb:116:in `block in <module:Renderers>'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/bundler/gems/active_model_serializers-e7d3323d2352/lib/action_controller/serialization.rb:51:in `block (2 levels) in <module:Serialization>'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/gems/actionpack-4.2.3/lib/action_controller/metal/renderers.rb:45:in `block in _render_to_body_with_renderer'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/gems/actionpack-4.2.3/lib/action_controller/metal/renderers.rb:41:in `_render_to_body_with_renderer'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/gems/actionpack-4.2.3/lib/action_controller/metal/renderers.rb:37:in `render_to_body'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/gems/actionpack-4.2.3/lib/abstract_controller/rendering.rb:25:in `render'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/gems/actionpack-4.2.3/lib/action_controller/metal/rendering.rb:16:in `render'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/gems/actionpack-4.2.3/lib/action_controller/metal/instrumentation.rb:44:in `block (2 levels) in render'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/gems/activesupport-4.2.3/lib/active_support/core_ext/benchmark.rb:12:in `block in ms'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/gems/activesupport-4.2.3/lib/active_support/core_ext/benchmark.rb:12:in `ms'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/gems/actionpack-4.2.3/lib/action_controller/metal/instrumentation.rb:44:in `block in render'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/gems/actionpack-4.2.3/lib/action_controller/metal/instrumentation.rb:87:in `cleanup_view_runtime'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/gems/activerecord-4.2.3/lib/active_record/railties/controller_runtime.rb:25:in `cleanup_view_runtime'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/gems/actionpack-4.2.3/lib/action_controller/metal/instrumentation.rb:43:in `render'
     # /Users/antonmurygin/.rvm/gems/ruby-2.2.2/gems/remotipart-1.2.1/lib/remotipart/render_overrides.rb:14:in `render_with_remotipart'
     # ./app/controllers/markers_controller.rb:20:in `block (2 levels) in local'
```

bf4 commented:
@kurko I'm trying to write a benchmark with positive results for caching but haven't been able to show differences with it on or off.  I'm probably missing something.  I see you had a working benchmark at one point. help?

https://github.com/rails-api/active_model_serializers/pull/810#issuecomment-89870165

```plain
Update: here are the numbers I got:

               user        system    total     real
no cache       21.550000   1.820000  23.370000 ( 28.894494)
cache          16.870000   1.580000  18.450000 ( 21.429540)
fragment cache 22.270000   1.810000  24.080000 ( 28.504920)

(cache means `only: []` wasn't used in the serializer)
```

https://github.com/rails-api/active_model_serializers/pull/810#issuecomment-94940858

```plain
# both fragment
cache only: [:field1, :field2, :etc]

# and not fragment
cache

# then, in an integration test:
Benchmark.bm do |x|
  x.report do
    1000.times do
      get "/users/#{user.id}", nil
    end
  end
end
```

beauby commented:
@bf4 Any change when changing [this line](https://github.com/rails-api/active_model_serializers/pull/810/files#diff-5624f9cf35be4dc254412d979fd293c4R48) from
```ruby
end unless method_defined?(key) || _fragmented.respond_to?(attr)
```
to
```ruby
end unless method_defined?(key) || _fragmented.respond_to?(key)
```
?

bf4 commented:
@beauby The number in the comment https://github.com/rails-api/active_model_serializers/pull/810#issuecomment-165063161 were a quote of @kurko .  I haven't been able to replicate much.. see #1393
